include/seastar/core/memory.hh

```diff
@@ -382,29 +383,76 @@ public:
     void operator=(scoped_large_allocation_warning_disable&&) = delete;
 };
 
-/// Enable/disable heap profiling.
+/// Describes an allocation location in the code.  The location is identified by
+/// its backtrace. One allocation_site can represent many allocations at the
+/// same location. `count` and `size` represent the cumulative sum of all
+/// allocations at the location.
+struct allocation_site {
+    mutable size_t count = 0; // number of live objects allocated at backtrace.
+    mutable size_t size = 0; // amount of bytes in live objects allocated at backtrace.
+    simple_backtrace backtrace; // call site for this allocation
+
+    // All allocation sites are linked to each other. This can be used for easy
+    // iteration across them in gdb scripts where it's difficult to work with
+    // the C++ data structures.
+    mutable const allocation_site* next = nullptr; // next allocation site in the chain
+    mutable const allocation_site* prev = nullptr; // previous allocation site in the chain
+
+    bool operator==(const allocation_site& o) const {
+        return backtrace == o.backtrace;
+    }
+
+    bool operator!=(const allocation_site& o) const {
+        return !(*this == o);
+    }
+};
+
+/// If memory sampling is on returns the current sampled memory live set
+std::vector<allocation_site> sampled_memory_profile();
+
+/// Copies the current sampled set of allocation_sites in the output parameter
+/// (up to size of output vector) Returns amount of copied elements. Useful if
+/// one wants to avoid allocating an output vector (e.g.: under OOM conditions)
+size_t sampled_memory_profile(std::vector<allocation_site>&);
+
+/// Enable sampled heap profiling by setting a sample rate
+/// disable heap profiling by setting the sample rate to 0
 ///
 /// In order to use heap profiling you have to define
 /// `SEASTAR_HEAPPROF`.
-/// Heap profiling data is not currently exposed via an API for
-/// inspection, instead it was designed to be inspected from a
-/// debugger.
+///
+/// Use \ref sampled_memory_profile for API access to profiling data
+///
 /// For an example script that makes use of the heap profiling data
 /// see [scylla-gdb.py] (https://github.com/scylladb/scylla/blob/e1b22b6a4c56b4f1d0adf65d1a11db4bcb51fe7d/scylla-gdb.py#L1439)
 /// This script can generate either textual representation of the data,
 /// or a zoomable flame graph ([flame graph generation instructions](https://github.com/scylladb/scylla/wiki/Seastar-heap-profiler),
 /// [example flame graph](https://user-images.githubusercontent.com/1389273/72920437-f0cf8a80-3d51-11ea-92f0-f3dbeb698871.png)).
-void set_heap_profiling_enabled(bool);
+void set_heap_profiling_sampling_rate(size_t);
+
+/// Checks whether heap profiling is currently enabled
+bool get_heap_profiling_enabled();
```

Comment:
Shouldn't this more usefully/symmetrically return the same `size_t` as passed to `set_heap_profiling_sampling_rate`?

Reply:
Brandon suggested to have a boolean version further above. I am happy to add another function.

Though maybe it's in the category of "we don't need it, lets see what seastar people want"?

Reply:
Minimal is good, I think the boolean one is redundant if you have the `size_t` one, so both is probably not the way to go. @ballard26 any specific reason it should be boolean on/off rather than the returning the sample rate since that's what the "set" side of the API works with?

Reply:
My thoughts on it are that `sampling_rate` and `enabled` are functionally independent configuration options. From that one should be able to change without effecting the other. And it's what I'm doing in the CPU profiler with a set of `get/set` methods for each.

That being said I'm not dead set on it if. And there is some sense to just have `sampling_rate == 0` entail that the profiler is disabled.

Reply:
Fair, have changed it back.